src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -203,15 +203,48 @@ public void AddToPendingBatch(ProvenBlockHeader provenBlockHeader, HashHeightPai
             if (this.saveAsyncLoopException != null)
                 throw this.saveAsyncLoopException;
 
+            List<KeyValuePair<int, ProvenBlockHeader>> itemsToRemove = null;
+
             lock (this.lockObject)
             {
+                if (this.pendingTipHashHeight != null && provenBlockHeader.HashPrevBlock != this.pendingTipHashHeight.Hash)
```

Comment:
technically it seems correct, but I would implement a simplier solution instead of this big if block:

```
foreach (int height in this.pendingBatch.Keys.Where(key => key >= newTip.Height).ToArray())
   this.pendingBatch.Remove(height);
```

basically we know that when we reorg, we are going backward, adding a lower height block in pendingBatch, so just removing all items from pendingBatch >= newTip.Height allow us to get rid of the old chain headers

Reply:
considering the block that remove the cache, it will be

```
if (this.pendingBatch.Count > 0)
{
    foreach (int height in this.pendingBatch.Keys.Where(key => key >= newTip.Height).ToArray()) { 
        this.pendingBatch.Remove(height);
        this.Cache.Remove(height);
    }
}
```

Reply:
Why iterate the entire collection?
Besides you can't to a remove inside a `foreach`

Reply:
sure, I do .toarray to take the snapshot, then iterating those methods, we are talking about very few items, I was preferring here readibility over performance becaue performance is not an issue here, but we can even do a single iteration with a store, from the last element to the first, checking if its key is >= newTip.height and in case remove the item at that position (and the cache too if it's fine to put it inside the lock)

Reply:
Here i preferred code readibility over performance optimization, we are talking about very few items, unnoticeable difference